### PR TITLE
fix #1083, ubuntu, diskless, enable the diskless installation log can be forwarded to the MN

### LINKED ISSUE #1083: [DEV] for diskless installation, the log can't forward to MN when xcatdebugmode=2
For redhat, sles, ubuntu.
The diskless installation log can't be forwarded to MN.

### PR DESCRIPTION
**Pull request by penguhyang**
fix #1083
For ubuntu diskless installation, enable the diskless installation log can be forwarded to the MN.

### CODE REVIEW

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 1]
```diff
@@ -1183,64 +1185,71 @@ ifconfig lo up
 
 XCATMASTER=`echo \$XCATSERVER|awk -F: '{print \$1}'`
 
+PORT="514"
+syslogd -R \$XCATMASTER:\$PORT
+
+[  "\$xcatdebugmode" = "1" ] || [  "\$xcatdebugmode" = "2" ] &&  logger -t xcat -p debug "running init script..."
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: should be ``([  "\$xcatdebugmode" = "1" ] || [  "\$xcatdebugmode" = "2" ]) &&  logger -t xcat -p debug "running init script..."``

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 2]
```diff
@@ -1183,64 +1185,71 @@ ifconfig lo up
...
+[  "\$xcatdebugmode" = "1" ] || [  "\$xcatdebugmode" = "2" ] &&  logger -t xcat -p debug "MASTER=\$XCATMASTER XCATIPORT=\$XCATIPORT"
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: same problem as above

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 3]
```diff
@@ -1183,64 +1185,71 @@ ifconfig lo up
...
 
 #update nodelist.nodestatus to "netbooting"
 if [ \$NODESTATUS != 'n' ]; then
+    [  "\$xcatdebugmode" = "1" ] || [  "\$xcatdebugmode" = "2" ] &&  logger -t xcat -p debug "nodestatus: netbooting,reporting..."
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: same problem as above

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 4]
```diff
@@ -1395,42 +1404,51 @@ fi
...
+    mount -t tmpfs rw /rw
+    mount -t aufs -o dirs=/rw:/ro mergedroot \$NEWROOT
+    mkdir -p \$NEWROOT/ro
+    mkdir -p \$NEWROOT/rw
+    mount --move /ro \$NEWROOT/ro
+    mount --move /rw \$NEWROOT/rw
 EOMS
    print $inifile "elif [ -r /rootimg.gz ]; then\n";
-   print $inifile "echo Setting up RAM-root tmpfs.\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"rootimg.gz downloaded,setting up RAM-root tmpfs...\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: should be ``([  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ]) && ``

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 5]
```diff
@@ -1395,42 +1404,51 @@ fi
...
-   print $inifile "  cd \$NEWROOT\n";
-   print $inifile "  echo -n \"Extracting root filesystem:\"\n";
-   print $inifile "  if [ -x /bin/cpio ]; then\n";
-   print $inifile "  zcat /rootimg.gz |/bin/cpio -idum\n";
-   print $inifile "  else\n";
-   print $inifile "  zcat /rootimg.gz |cpio -idum\n";
-   print $inifile "  fi\n";
-   print $inifile "  echo Done\n";
+   print $inifile "    cd \$NEWROOT\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"Extracting root filesystem:\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic problem

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 6]
```diff
@@ -1395,42 +1404,51 @@ fi
...
+   print $inifile "    echo -n \"Extracting root filesystem:\"\n";
+   print $inifile "    if [ -x /bin/cpio ]; then\n";
+   print $inifile "        zcat /rootimg.gz |/bin/cpio -idum\n";
+   print $inifile "    else\n";
+   print $inifile "        zcat /rootimg.gz |cpio -idum\n";
+   print $inifile "    fi\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"Done...\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic problem

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 7]
```diff
@@ -1395,42 +1404,51 @@ fi
...
+   print $inifile "    echo Done\n";
    print $inifile "else\n";
-   print $inifile "  echo -n Failed to download image, panicing in 5...\n";
-   print $inifile "  for i in 4 3 2 1 0; do\n";
-   print $inifile "    /bin/sleep 5\n";
-   print $inifile "    echo -n \$i...\n";
-   print $inifile "  done\n";
-   print $inifile "  echo\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"Failed to download image, panicing in 5...\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic problem

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 8]
```diff
@@ -1395,42 +1404,51 @@ fi
...
+   print $inifile "    echo -n Failed to download image, panicing in 5...\n";
+   print $inifile "    for i in 4 3 2 1 0; do\n";
+   print $inifile "        /bin/sleep 5\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"\$i...\"\n";
+   print $inifile "        echo -n \$i...\n";
+   print $inifile "    done\n";
+   print $inifile "    [  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"You're dead.  rpower nodename reset to play again.\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic problem

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 9]
```diff
@@ -1440,13 +1458,15 @@ EOMS
...
 "
-  sleep 5
+    sleep 5
 EOMS
-   print $inifile "  exit\n";
+   print $inifile "    exit\n";
    print $inifile "fi\n";
    print $inifile "cd /\n";
    print $inifile "cp /etc/hostname \$NEWROOT/etc/hostname\n";
+   print $inifile "[  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"saving \$NEWROOT/etc/resolv.conf\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic error

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 10]
```diff
@@ -1440,13 +1458,15 @@ EOMS
...
    print $inifile "rm -f \$NEWROOT/etc/resolv.conf; cp /etc/resolv.conf \$NEWROOT/etc/resolv.conf\n";
+   print $inifile "[  \"\$xcatdebugmode\" = \"1\" ] || [  \"\$xcatdebugmode\" = \"2\" ] &&  logger -t xcat -p debug \"exiting init script...\"\n";
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: logic error

**Comment by penguhyang** [comment]
@immarvin Thanks for review.
I will add parenthesis to make more clear.

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 11]
```diff
@@ -1183,64 +1185,71 @@ ifconfig lo up
 
 XCATMASTER=`echo \$XCATSERVER|awk -F: '{print \$1}'`
 
+PORT="514"
+syslogd -R \$XCATMASTER:\$PORT
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: rsyslog should only be started when XCATDEBUGMODE>=1

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 12]
```diff
@@ -1183,64 +1185,73 @@ ifconfig lo up
 
 XCATMASTER=`echo \$XCATSERVER|awk -F: '{print \$1}'`
 
+if [ "\$xcatdebugmode" = "1" ] || [ "\$xcatdebugmode" = "2" ]; then
```
Comment (penguhyang) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: @immarvin Thanks for review.
Add judgement whether rsyslog should be started.

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 13]
```diff
@@ -1183,64 +1185,73 @@ ifconfig lo up
...
+    elif [ "\$KEY" == 'NFSROOT' ]; then
+        NFSROOT=1
+        VALUE=`echo \$i |awk -F= '{print \$2}'`
+        SERVER=`echo \$VALUE|awk -F: '{print \$1}'`
+        ROOTDIR=`echo \$VALUE|awk -F/ '{for(i=2;i<=NF;i++) printf "/%s",\$i}'`
+    elif [ "\$KEY" == 'STATEMNT' ]; then
+        STATELITE=1
+        VALUE=`echo \$i |awk -F= '{print \$2}'`
+        # the VALUE may be null
+        if [ ! -z \$VALUE ]; then
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: ``"\$VALUE"``

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 14]
```diff
@@ -1183,64 +1185,73 @@ ifconfig lo up
...
+            SNAPSHOTSERVER=`echo \$VALUE|awk -F: '{print \$1}'`
+            SNAPSHOTROOT=`echo \$VALUE|awk -F/ '{for(i=2;i<=NF;i++) printf "/%s",\$i}'`
+            # may be that there is not server and just a directory.
+            if [ -z \$SNAPSHOTROOT ]; then
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: ``"\$SNAPSHOTROOT"``

File: xCAT-server/share/xcat/netboot/ubuntu/genimage (hunk 13 above)
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: all the string compare/test inside ``[]`` should be double quoted

Comment (penguhyang) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: fixed

File: xCAT-server/share/xcat/netboot/ubuntu/genimage (hunk 14 above)
Comment (penguhyang) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: fixed

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 15]
```diff
@@ -1145,13 +1145,15 @@ for i in `cat /proc/cmdline`; do
         VALUE=`echo \$i |awk -F= '{print \$2}'`
         # format: XCAT=xcatmaster:3001
         XCATIPORT=\$VALUE
+    elif [ "\$KEY" == 'xcatdebugmode' ]; then
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: in string comparison in ``[]``, the equal  should be ``=``;
``==`` is used in ``[[]]``

File: xCAT-server/share/xcat/netboot/ubuntu/genimage [hunk 16]
```diff
@@ -1145,13 +1145,15 @@ for i in `cat /proc/cmdline`; do
...
+        VALUE=`echo \$i |awk -F= '{print \$2}'`
+        xcatdebugmode=\$VALUE
     fi
     
     #if "nonodestatus" specified,do not update the nodestatus 
     if [ \$i == 'nonodestatus' ]; then
```
Comment (immarvin) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: should be ``"\$i"``

Comment (penguhyang) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: fixed.

File: xCAT-server/share/xcat/netboot/ubuntu/genimage (hunk 15 above)
Comment (penguhyang) [inline xCAT-server/share/xcat/netboot/ubuntu/genimage]: fixed.

**Comment by immarvin** [comment]
agree to merge

**Pull request merged**